wgsl/extract-grammar.py

```
@@ -255,118 +338,287 @@ def parse(lines, i):
+                    [reproduce_rule_source(member) for member in production["value"]]
+                )
+        elif production["type"] == "choice":
+            return (
+                "( "
+                + " | ".join(
+                    [reproduce_rule_source(member) for member in production["value"]]
+                )
+                + " )"
+            )
+        elif production["type"] == "match_0_1":
+            if len(production["value"]) == 1:
+                return reproduce_rule_source(production["value"][0]) + " ?"
+            else:
+                return (
+                    "( "
+                    + " ".join(
+                        [
+                            reproduce_rule_source(member)
+                            for member in production["value"]
+                        ]
+                    )
+                    + " ) ?"
+                )
+        elif production["type"] == "match_1_n":
+            if len(production["value"]) == 1:
+                return reproduce_rule_source(production["value"][0]) + " +"
+            else:
+                return (
+                    "( "
+                    + " ".join(
+                        [
+                            reproduce_rule_source(member)
+                            for member in production["value"]
+                        ]
+                    )
+                    + " ) +"
+                )
+        elif production["type"] == "match_0_n":
+            if len(production["value"]) == 1:
+                return reproduce_rule_source(production["value"][0]) + " *"
+            else:
+                return (
+                    "( "
+                    + " ".join(
+                        [
+                            reproduce_rule_source(member)
+                            for member in production["value"]
+                        ]
+                    )
+                    + " ) *"
+                )
+        else:
+            print(production["type"])
```

Should this raise an exception?  It happens when there's a missing case in this method.
Also, it does not have to be in the "else" clause.

---

That is correct